vhost-device-input/src/main.rs

```diff
@@ -0,0 +1,226 @@
+        })
+    }
+}
+
+pub(crate) fn start_backend(config: VuInputConfig) -> Result<()> {
+    let mut handles = Vec::new();
+
+    for i in 0..config.count {
+        let socket = format!("{}{}", config.socket_path.to_owned(), i);
+        let event = config.events.get(i as usize).unwrap().clone();
+
+        let handle: JoinHandle<Result<()>> = thread::spawn(move || loop {
+            let ev_dev = match evdev::Device::open(&event) {
+                Ok(dev) => dev,
+                Err(err) => {
+                    error!("Failed to open event device {}, error is: {}", event, err);
+                    return Err(Error::AccessEventDeviceFile);
+                }
+            };
+
+            let raw_fd = ev_dev.as_raw_fd();
+
+            // If creating the VuInputBackend isn't successull there isn't much else to do than
+            // killing the thread, which .unwrap() does.  When that happens an error code is
+            // generated and displayed by the runtime mechanic.  Killing a thread doesn't affect
+            // the other threads spun-off by the daemon.
+            let vu_input_backend = Arc::new(RwLock::new(
+                VuInputBackend::new(ev_dev).map_err(Error::CouldNotCreateBackend)?,
+            ));
+
+            let mut daemon = VhostUserDaemon::new(
+                String::from("vhost-device-input-backend"),
+                Arc::clone(&vu_input_backend),
+                GuestMemoryAtomic::new(GuestMemoryMmap::new()),
+            )
+            .map_err(Error::CouldNotCreateDaemon)?;
+
+            let handlers = daemon.get_epoll_handlers();
+            handlers[0]
+                .register_listener(raw_fd, EventSet::IN, vhu_input::EVENT_ID_IN_VRING_EPOLL)
+                .map_err(|_| Error::CouldNotRegisterInputEvent)?;
+
+            let listener = Listener::new(socket.clone(), true).unwrap();
+            let _ = daemon.start(listener).map_err(|_| Error::ServeFailed);
+
+            match daemon.wait() {
+                Ok(()) => {
+                    info!("Stopping cleanly.");
+                }
+                Err(vhost_user_backend::Error::HandleRequest(
+                    vhost_user::Error::PartialMessage | vhost_user::Error::Disconnected,
+                )) => {
+                    info!("vhost-user connection closed with partial message. If the VM is shutting down, this is expected behavior; otherwise, it might be a bug.");
+                }
+                Err(e) => {
+                    warn!("Error running daemon: {:?}", e);
+                }
+            }
+
+            // No matter the result, we need to shut down the worker thread.
+            vu_input_backend
+                .read()
+                .unwrap()
+                .exit_event
+                .write(1)
+                .expect("Shutting down worker thread");
+        });
+
+        handles.push(handle);
+    }
+
+    for handle in handles {
+        handle.join().map_err(|_| Error::FailedJoiningThreads)??;
+    }
```

Comment:
Good idea.  I tried to use thread::scope(), but in the end it requires to conver the Error type for io::Error:

```
impl convert::From<Error> for io::Error {
    fn from(e: Error) -> Self {
        io::Error::new(io::ErrorKind::Other, e)
    }
}
```

When built this section code, rustc reports error:

```
  Compiling vhost-device-input v0.1.0 (/home/leoy/Dev1/orko/vhost-device/vhost-device-input)
error[E0277]: `(dyn Any + Send + 'static)` cannot be shared between threads safely
   --> vhost-device-input/src/main.rs:47:46
    |
47  |         io::Error::new(io::ErrorKind::Other, e)
    |         --------------                       ^ `(dyn Any + Send + 'static)` cannot be shared between threads safely
    |         |
    |         required by a bound introduced by this call
    |
    = help: the trait `Sync` is not implemented for `(dyn Any + Send + 'static)`
    = note: required for `Unique<(dyn Any + Send + 'static)>` to implement `Sync`
note: required because it appears within the type `Box<dyn Any + Send>`
   --> /home/leoy/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:195:12
    |
195 | pub struct Box<
    |            ^^^
note: required because it appears within the type `Error`
   --> /home/leoy/.cargo/registry/src/index.crates.io-6f17d22bba15001f/vhost-user-backend-0.11.0/src/lib.rs:41:10
    |
41  | pub enum Error {
    |          ^^^^^
note: required because it appears within the type `Error`
   --> vhost-device-input/src/main.rs:28:17
    |
28  | pub(crate) enum Error {
    |                 ^^^^^
    = note: required for `Box<(dyn std::error::Error + Send + Sync + 'static)>` to implement `From<Error>`
    = note: required for `Error` to implement `Into<Box<(dyn std::error::Error + Send + Sync + 'static)>>`
note: required by a bound in `std::io::Error::new`
   --> /home/leoy/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/io/error.rs:516:12
    |
514 |     pub fn new<E>(kind: ErrorKind, error: E) -> Error
    |            --- required by a bound in this associated function
515 |     where
516 |         E: Into<Box<dyn error::Error + Send + Sync>>,
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `Error::new`
```

Will keep this comment as opening.   I will address other comments firstly and come back to check if I can refine this part :)